### PR TITLE
add combined derived attributes

### PR DESCRIPTION
**Pull request by pordyna**
# Combined derived attributes

This PR introduces derived attributes which are a function of two primary derived attributes (such like we had until now). This is something I need for my online Faraday Rotation calculation but I found this a generally useful feature. 

What I needed was 1/gamma^2 * density, where gamma is the Lorentz factor for the average kinetic energy in the cell. Because of the square it wasn't possible to just add another usual derived field. This new combined field is introduced as `RelativisticDensity`. 

This PR also includes a template for an average value of any derived quantity `AverageAttribute` ( from: average value = total quantity density / number density) as an example of a more general use of this feature. 

I also changed how the `computeValue` algorithm is called in the `openPMD` and `isaac` plugins so that the same syntax can be used for computing a normal and combined derived attribute. 

I think in most cases writing both fields individually would be a simpler solution but when fields are written or streamed for a large number of iterations and output size starts to really matter this could be a great help (definitely would be in my Faraday Rotation simulations that I was running earlier this year).  And it can be used internally , wherever a more general solution is needed (like in my `Scatterer` stage implelemtation). 

I tested this for  `AverageAttribute<Energy>` with the `FoilLTC` example and could verify that it is same as writing `EnergyDensity` and `Denisty` individually and dividing the first by the second (and setting to 0 wherever Density is 0).

@ComputationalRadiationPhysics/picongpu-maintainers

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
I will stop the CI for your job, you need to rebase against #3942 or wait until #3942 is merged into the dev branch, else the CI will fail.

**Comment by pordyna** [comment]
Ok, I will just wait for #3942.

**Comment by psychocoderHPC** [comment]
Please rebase against the current dev branch to pull the latest fixes to be able to pass the CI.

**Comment by pordyna** [comment]
done :)

File: include/picongpu/particles/particleToGrid/ComputeFieldValue.hpp
Comment (PrometheusPi) [inline include/picongpu/particles/particleToGrid/ComputeFieldValue.hpp:121]:
```suggestion
                    /* Particles can contribute to cells in GUARD (due to their shape) these values need to be
```

File: include/picongpu/particles/particleToGrid/combinedAttributes/AverageAttribute.hpp
Comment (PrometheusPi) [inline include/picongpu/particles/particleToGrid/combinedAttributes/AverageAttribute.hpp:42]: How do you avoid close to infinite results caused by division of barely zero (e.g.  if you are in a vacuum boarder where macro-particles in neighboring cells just contribute very little due to their shape and thus you divide by a ver y small number?)

File: include/picongpu/plugins/IsaacPlugin.hpp
Comment (PrometheusPi) [inline include/picongpu/plugins/IsaacPlugin.hpp:137]: Missing what you need to allocate (one or two of what?)

**Review by PrometheusPi — COMMENTED** [review]
Overall looks really good to me - just a few remarks and questions

File: include/picongpu/particles/particleToGrid/combinedAttributes/AverageAttribute.hpp
Comment (pordyna) [reply include/picongpu/particles/particleToGrid/combinedAttributes/AverageAttribute.hpp:42]: @PrometheusPi  Well, i hope this won't happen. At least, I didn't see it happen. Whenever density is very small the total quantity should also be very small (it scales with weighting as well) so they should cancel out. 

Btw. it probably should be:
```suggestion
                    if(dens * particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE * CELL_VOLUME 
                        <= std::numeric_limits<float_X>::min())
```

Maybe the limit should also be a bit higher? One could even go with `< 1.0` and set to zero for all cells with the number of real particles below 1 and just properly document this behavior. But I guess that would be some lost information.

File: include/picongpu/plugins/IsaacPlugin.hpp
Comment (pordyna) [reply include/picongpu/plugins/IsaacPlugin.hpp:137]:
```suggestion
                        _please_allocate_at_least_one_FieldTmp_slot_in_memory_param_or_two_when_using_combined_attributes,
```

Comment (PrometheusPi) [reply include/picongpu/plugins/IsaacPlugin.hpp:137]: I like that error message.

File: include/picongpu/particles/particleToGrid/combinedAttributes/AverageAttribute.hpp
Comment (PrometheusPi) [reply include/picongpu/particles/particleToGrid/combinedAttributes/AverageAttribute.hpp:42]: I think you are right - it should cancel out as long as the power of numerator is not `< 1`. 

I like the `< 1.0` limit, but I agree with you that this might become some lost information and additionally might cause some strange errors for simulations with weightings below 1 (2D TNSA simulations). 

I would think keeping it for now and seeing, whether issues arise at vacuum boarders, might be a more practical approach. I assume you will use this methods extensively in the future for simulations where there is also vacuum involved, correct? Then you will  see whether issues arise and can fix it in a followup pull request.

**Review by PrometheusPi — DISMISSED** [review]

**Comment by pordyna** [comment]
@PrometheusPi Thanks for your review! I just added this suggestion https://github.com/ComputationalRadiationPhysics/picongpu/pull/3943#discussion_r772976773

**Review by PrometheusPi — APPROVED** [review]
Thanks for implementing the suggestions.

**Review by psychocoderHPC — APPROVED** [review]

**Comment by PrometheusPi** [comment]
@psychocoderHPC Do we actually perform compile test with ISAAC in PIConGPU? 
When compiling ISAAC with the current dev I run into the following compile error:
```
In file included from /.../picongpu/include/picongpu/main.cpp:23:
In file included from /.../picongpu/include/pmacc/../picongpu/simulation_defines.hpp:55:
In file included from /.../picongpu/include/pmacc/../picongpu/unitless/starter.unitless:23:
In file included from /.../picongpu/include/pmacc/../picongpu/plugins/PluginController.hpp:69:
/.../picongpu/include/pmacc/../picongpu/plugins/IsaacPlugin.hpp:142:25: error: no member named 'ComputeFieldValue' in namespace 'picongpu::particles::particleToGrid'
                        ComputeFieldValue<CORE + BORDER, FrameSolver, ParticleType, ParticleFilter>()
```

**Pull request merged**
